Allow admin access to datasets API

davelopez opened the pull request:
Some of the endpoints in the datasets API were using the [get_hda_or_ldda](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/webapps/base/controller.py#L984) function to retrieve the dataset association. Since this code is around 10 years old I figured out that we probably should use the existing manager's methods instead which also take into account Admin access and unify the error handling.

This should take care of the other API related issue in https://github.com/galaxyproject/galaxy/issues/13001#issuecomment-981717424

## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

davelopez committed:
Add API test

Admins should be able to change datasets permissions

davelopez committed:
Fix admin access to various datasets API endpoints

- Replace the old `get_hda_or_ldda` function to retrieve datasets with the `get_accessible` function from the appropriate manager. This takes into account admin access and unifies error handling.

davelopez committed:
Adapt `test_show_permission_denied` to expect 403 status code

mvdbeek reviewed (approved):
Very nice, thank you! Do you want to target 21.09 ?

davelopez committed:
Adapt API test to expect 403 status code

After aplying the fix #13056, `test_sharing_private_history_makes_datasets_public` will now return a proper 403 status code instead of 400.
